### ISSUE TITLE
when using apache2.4 with mod_gssapi I need to turn set KRB5RCACHETYPE=none

### ISSUE THREAD

**Issue by natxo** [jcmturner/gokrb5#170]
hi,

When using a simple http client with spnego headers against a centos 7 with apache2.4 and mod_gssapi, I need to turn off that variable in the server.

I found this in the error_log of apache:

```
GSS ERROR In Negotiate Auth: gss_accept_sec_context() failed: [Unspecified GSS failure.  Minor c
ode may provide more information (Request is a replay)], referer: http://host.sub.domain.tld/private
````
See https://github.com/modauthgssapi/mod_auth_gssapi/issues/83 for the details

This is the go code:

```
package main

import (
	"fmt"
	"gopkg.in/jcmturner/gokrb5.v5/client"
	"gopkg.in/jcmturner/gokrb5.v5/config"
	"net/http"
)

func main() {
	c, err := config.Load("/etc/krb5.conf")
	if err != nil {
		panic(err)
	}
	c.LibDefaults.DNSLookupKDC = true
	cl := client.NewClientWithPassword("user", "SUB.DOMAIN.TLD", "pwd")
	cl.WithConfig(c)
	err = cl.Login()
	if err != nil {
		panic(err)
	}

	r, _ := http.NewRequest("GET", "http://host.sub.domain.tld/private", nil)

	spn := ""
	cl.SetSPNEGOHeader(r, spn)
	HTTResp, err := http.DefaultClient.Do(r)
	if err != nil {
		panic(err)
	}

	fmt.Println(HTTResp)

}

```
If I disable that environment variable, then this is the result:

`&{401 Unauthorized 401 HTTP/1.1 1 1 map[Content-Length:[381] Content-Type:[text/html; charset=iso-8859-1] Date:[Thu, 12 Jul 2018 19:35:48 GMT] Server:[Apache/2.4.6 (CentOS) mod_auth_gssapi/1.5.1] Www-Authenticate:[Negotiate oQcwBaADCgEC]] 0xc4200a2580 381 [] false false map[] 0xc420198000 <nil>}`

**Comment by natxo** [comment]
just for completeness, curl --negotiate --user : http://host.sub.domain.tld/private works correctly

**Comment by jcmturner** [comment]
@natxo for the first error that references a replay are you making the request to the server more than once from the client? The code you have posted seems to imply not but the error you are getting is usually when a client has tried to use the same ticket twice.

For the second error do you have any logs from the apache server for  this as I think this will help diagnose what the issue is here. Thanks.

**Comment by jcmturner** [comment]
@natxo do you have details on your apache 2.4 config so that I can look to recreate this for testing?

**Comment by natxo** [comment]
@jcmturner ,
apologies for the delay in getting back to you, life happened.

This is the relevant httpd config:
<Location /private>
    AuthType GSSAPI
    AuthName "GSSAPI Single Sign On Login"
    GssapiCredStore keytab:/etc/httpd/conf.d/fqdn.keytab
    Require valid-user
</Location>

The host is running centos 7 with the default apache2 and mod_auth_gssapi like indicated here: https://access.redhat.com/articles/3161491 or here: https://fedorapeople.org/cgit/adelton/public_git/CGI-sessions.git/commit/?h=negotiate-mod_auth_gssapi

**Comment by jcmturner** [comment]
@natxo I've tried to recreate your issue but I've been unable to.

You can see that the integration test below tests against mod_auth_gssapi:
https://github.com/jcmturner/gokrb5/blob/2bdd8614e51c306c067b105027d299f42852262f/client/client_integration_test.go#L321

The web server being tested against is using mod_auth_gssapi configured as you have it:
https://github.com/jcmturner/gokrb5/blob/2bdd8614e51c306c067b105027d299f42852262f/testenv/docker/krbhttp/httpd-krb5.conf#L3-L8

The tests pass and therefore I am unable to recreate your issue.

I notice in the code you provided above you are not checking the error return from the ``SetSPNEGOHeader`` function. Perhaps check that and see if there is some other issue going on here?

**Comment by natxo** [comment]
just tried this:
```
//cl.SetSPNEGOHeader(r, spn)
 30     err = cl.SetSPNEGOHeader(r, spn)                    
 31     if err != nil {                         
 32         panic(err)                                             
 33     }  
```
When running it and capturing the traffic in Wireshark, I clearly see negotiate header in the packet.

The difference in the traffic capture is that (using the kerberos filter) the go client does a GET request twice and a 401 comes back. 

The curl client does first a GET, the server sends a 301 back, the curl client requests the GET again (same url, then the server sends the document with 200 OK.

Unsure as to why

**Comment by jcmturner** [comment]
@natxo when using the go client is the first response back from the server a 301 and then after the second request you get a 401? That is, is the sequence?...
```
go client                 server
     |------GET------>|
     |<-----301-------| 
     |------GET------>|
     |<-----401-------|
```

**Comment by natxo** [comment]
@jcmturner indeed, yes, first 301, then 401. I could send you a pcap if you wanted

**Comment by jcmturner** [comment]
OK I think I'm understanding this a bit more now. The URL you are first visiting is sending a redirect. The gokrb5 client is sending the kerberos ticket on the first request. The web server is adding this to its cache and sending the redirect. The Go http client automatically follows the redirect sending the same header and kerberos ticket again. The web server therefore thinks this is a replay and returns a 401.

I need to figure out how to handle the automatic behaviour of the Go http client.

In your capture does the web server set a cookie? The pcap may be helpful to see this.

**Comment by jcmturner** [comment]
@natxo I have released the spngeo.Client in v7.1.0

Let me know if this resolve things.

**Comment by natxo** [comment]
@jcmturner ,
after some trial and error (I needed to import the spnego library or I would get an undefined spnego error and the password client required an extra argument, so I needed to modify those settings too):

```

package main

import (
	"fmt"
	"gopkg.in/jcmturner/gokrb5.v7/client"
	"gopkg.in/jcmturner/gokrb5.v7/config"
	"gopkg.in/jcmturner/gokrb5.v7/config"	
        "net/http"
)

func main() {
	c, err := config.Load("/etc/krb5.conf")
	if err != nil {
		panic(err)
	}
	c.LibDefaults.DNSLookupKDC = true
	cl := client.NewClientWithPassword("user", "SUB.DOMAIN.TLD", "pwd", c)
	err = cl.Login()
	if err != nil {
		panic(err)
	}
	fmt.Print(cl)

	r, err := http.NewRequest("GET", "http://host.sub.domain.tld/private", nil)
	if err != nil {
		panic(err)
	}

	spnegocl := spnego.NewClient(cl, nil, "")
	resp, err := spnegocl.Do(r)

	fmt.Println(resp)

}
```

but:

$ go run test.go 
&{0xc0001b00c0 0xc0000fc000 0xc000011630 0xc00000d580 0xc00000d560}&{401 Unauthorized 401 HTTP/1.1 1 1 map[Server:[Apache/2.4.6 (CentOS) mod_auth_gssapi/1.5.1] Www-Authenticate:[Negotiate] Content-Length:[381] Content-Type:[text/html; charset=iso-8859-1] Date:[Sat, 09 Feb 2019 16:25:22 GMT]] 0xc0002b0040 381 [] false false map[] 0xc0002d8300 <nil>}

So now with the variable KRB5RCACHETYPE=none  I cannot log in.

**Comment by jcmturner** [comment]
Does the err variable from the spnegocl.Do have any information? Could you test if it's nil and if not print it out.

**Comment by natxo** [comment]
&{0xc0001c20c0 0xc0000fc000 0xc000011630 0xc00000d580 0xc00000d560}panic: nil

goroutine 1 [running]:
main.main()
	/home/user/gokrb/test.go:35 +0x262
exit status 2

**Comment by jcmturner** [comment]
@natxo I assume that line 35 is the panic inside and if statement checking the err from the spnego.Do?

Are you able to see from tcpdump/wireshark what the flow of interaction is with the web server and what the headers are for these messages? This will help understand what is going on a bit more.

Are you using apache to reverse proxy to another site or application server? Could this be sending the 401 and it is being proxied back via the apache server?

**Comment by natxo** [comment]
@jcmturner ,
no reverse proxy involved.
Could I send you a pcap? I'd rather not post it here, so if you could give me an e-mail address/ or send me one to nasenjo_remove_this@removethis_asenjo.nl I'll send it to you.
